README.md
# ExAws.Lambda

[![Module Version](https://img.shields.io/hexpm/v/ex_aws_lambda.svg)](https://hex.pm/packages/ex_aws_lambda)
[![Hex Docs](https://img.shields.io/badge/hex-docs-lightgreen.svg)](https://hexdocs.pm/ex_aws_lambda/)
[![Total Download](https://img.shields.io/hexpm/dt/ex_aws_lambda.svg)](https://hex.pm/packages/ex_aws_lambda)
[![License](https://img.shields.io/hexpm/l/ex_aws_lambda.svg)](https://github.com/ex-aws/ex_aws_lambda/blob/master/LICENSE)
[![Last Updated](https://img.shields.io/github/last-commit/ex-aws/ex_aws_lambda.svg)](https://github.com/ex-aws/ex_aws_lambda/commits/master)

Service module for [https://github.com/ex-aws/ex_aws_lambda](https://github.com/ex-aws/ex_aws_lambda).

## Installation

The package can be installed by adding `:ex_aws_lambda` to your list of dependencies in `mix.exs`
along with `:ex_aws` and your preferred JSON codec / HTTP client

```elixir
def deps do
  [
    {:ex_aws, "~> 2.0"},
    {:ex_aws_lambda, "~> 2.0"},
    {:poison, "~> 3.0"},
    {:hackney, "~> 1.9"},
  ]
end
```

Documentation can be found at [https://hexdocs.pm/ex_aws_lambda](https://hexdocs.pm/ex_aws_lambda).

## License

The MIT License (MIT)

Copyright (c) 2014 CargoSense, Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## v2.2.0
* BREAKING: Bump minimum elixir version to 1.11

## v2.1.0 - 2022-09-02

* Update minimum Elixir version to 1.7
* Add dependabot and github workflow
* Fix endpoint for update function configuration operation
* Add operation to tag a resource

## v2.0.2 - 2020-12-22

* Add `trace_id` as `X-Amzn-Trace-Id` if provided in options

## v2.0.1 - 2018-11-17

* Update library configuration

## v2.0.0 - 2017-11-10

* Major Project Split. Please see the main ExAws repository for previous changelogs.


test/lib/lambda_test.exs
defmodule ExAws.LambdaTest do
  use ExUnit.Case, async: true

  test "#list_functions" do
    assert %ExAws.Operation.JSON{
             before_request: nil,
             data: %{},
             headers: [{"content-type", "application/json"}],
             http_method: :get,
             params: %{},
             parser: _,
             path: "/2015-03-31/functions/?",
             service: :lambda,
             stream_builder: nil
           } = ExAws.Lambda.list_functions()
  end

  describe "Lambda.invoke" do
    test "callback on invoke works for encoding context before request" do
      op = ExAws.Lambda.invoke("func-name", "func-payload", "asdfasdf")
      before_request = op.before_request.(op, %{json_codec: Poison})

      assert {"X-Amz-Client-Context", "ImFzZGZhc2RmIg=="} in before_request.headers
    end

    test "xray_trace_id option adds the X-Amzn-Trace-Id header" do
      op =
        ExAws.Lambda.invoke("func-name", "func-payload", %{}, xray_trace_id: "1-aaaaa-bbbbbbbbb")

      assert {"X-Amzn-Trace-Id", "1-aaaaa-bbbbbbbbb"} in op.headers
    end

    test "builds ExAws JSON operation" do
      assert %ExAws.Operation.JSON{
               before_request: _,
               data: "func-payload",
               headers: [
                 {"content-type", "application/json"}
               ],
               http_method: :post,
               params: %{},
               parser: _,
               path: "/2015-03-31/functions/func-name/invocations?",
               service: :lambda,
               stream_builder: nil
             } = ExAws.Lambda.invoke("func-name", "func-payload", %{})
    end

    test "builds ExAws RestQuery operation" do
      assert %ExAws.Operation.RestQuery{
               body: "func-payload",
               http_method: :post,
               params: %{},
               parser: parser_func,
               path: "/2015-03-31/functions/func-name/invocations?",
               service: :lambda
             } = ExAws.Lambda.invoke("func-name", "func-payload", %{}, parser: &{&1, &2})

      assert is_function(parser_func, 2)
    end
  end
end


mix.exs
defmodule ExAws.Lambda.Mixfile do
  use Mix.Project

  @version "2.2.0"
  @service "lambda"
  @url "https://github.com/ex-aws/ex_aws_#{@service}"
  @name __MODULE__ |> Module.split() |> Enum.take(2) |> Enum.join(".")

  def project do
    [
      app: :ex_aws_lambda,
      version: @version,
      elixir: "~> 1.11",
      elixirc_paths: elixirc_paths(Mix.env()),
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      name: @name,
      package: package(),
      docs: docs()
    ]
  end

  defp package do
    [
      description: "#{@name} service package",
      files: ["lib", "config", "mix.exs", "README*", "CHANGELOG*"],
      maintainers: ["Bernard Duggan"],
      licenses: ["MIT"],
      links: %{github: @url}
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:hackney, ">= 0.0.0", only: [:dev, :test]},
      {:sweet_xml, ">= 0.0.0", only: [:dev, :test]},
      {:poison, ">= 0.0.0", only: [:dev, :test]},
      {:dialyxir, "~> 1.0", only: [:dev, :test], runtime: false},
      {:ex_doc, ">= 0.0.0", only: [:dev], runtime: false},
      ex_aws()
    ]
  end

  defp docs do
    [
      extras: ["CHANGELOG.md", "CONTRIBUTING.md", "README.md"],
      main: "readme",
      source_ref: "v#{@version}",
      source_url: @url,
      formatters: ["html"]
    ]
  end

  defp ex_aws() do
    case System.get_env("AWS") do
      "LOCAL" -> {:ex_aws, path: "../ex_aws"}
      _ -> {:ex_aws, "~> 2.0"}
    end
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure your application as:
#
#     config :ex_aws_lambda, key: :value
#
# and access this configuration in your application as:
#
#     Application.get_env(:ex_aws_lambda, :key)
#
# You can also configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/ex_aws/lambda.ex
defmodule ExAws.Lambda do
  @moduledoc """
  Operations on ExAws Lambda.
  """

  import ExAws.Utils, only: [camelize_key: 1, camelize_keys: 1, upcase: 1]
  require Logger

  @actions %{
    add_permission: :post,
    create_event_source_mapping: :post,
    create_function: :post,
    delete_event_source_mapping: :delete,
    delete_function: :delete,
    get_event_source_mapping: :get,
    get_function: :get,
    get_function_configuration: :get,
    get_policy: :get,
    invoke: :post,
    invoke_async: :post,
    list_event_source_mappings: :get,
    list_functions: :get,
    remove_permission: :delete,
    update_event_source_mapping: :put,
    update_function_code: :put,
    update_function_configuration: :put,
    tag_resource: :post
  }

  @type starting_position_vals :: :trim_horizon | :latest

  @type add_permission_opts :: [
          {:source_account, binary}
          | {:source_arn, binary}
        ]

  @doc """
  Adds a permission to the access policy associated with the specified AWS Lambda function.

  Action pattern: (lambda:[*]|lambda:[a-zA-Z]+|[*])
  """
  @spec add_permission(
          function_name :: binary,
          principal :: binary,
          action :: binary,
          statement_id :: binary
        ) :: ExAws.Operation.JSON.t()
  @spec add_permission(
          function_name :: binary,
          principal :: binary,
          action :: binary,
          statement_id :: binary,
          opts :: add_permission_opts
        ) :: ExAws.Operation.JSON.t()
  def add_permission(function_name, principal, action, statement_id, opts \\ []) do
    data =
      opts
      |> normalize_opts
      |> Map.merge(%{
        "Action" => action,
        "Principal" => principal,
        "StatementId" => statement_id
      })

    request(:add_permission, data, "/2015-03-31/functions/#{function_name}/versions/HEAD/policy")
  end

  @type create_event_source_mapping_opts :: [
          {:batch_size, pos_integer}
          | {:enabled, boolean}
        ]
  @doc """
  Creates a stream based event source for a function.
  """
  @spec create_event_source_mapping(
          function_name :: binary,
          event_source_arn :: binary,
          starting_position :: starting_position_vals
        ) :: ExAws.Operation.JSON.t()
  @spec create_event_source_mapping(
          function_name :: binary,
          event_source_arn :: binary,
          starting_position :: starting_position_vals,
          opts :: create_event_source_mapping_opts
        ) :: ExAws.Operation.JSON.t()
  def create_event_source_mapping(function_name, event_source_arn, starting_position, opts \\ []) do
    data =
      opts
      |> normalize_opts
      |> Map.merge(%{
        "FunctionName" => function_name,
        "EventSourceArn" => event_source_arn,
        "StartingPosition" => starting_position |> upcase
      })

    request(:create_event_source_mapping, data, "/2015-03-31/event-source-mappings/")
  end

  @type create_function_opts :: [
          {:description, binary}
          | {:memory_size, pos_integer}
          | {:timeout, pos_integer}
        ]
  @doc """
  Create a function.

  Runtime defaults to nodejs, as that is the only one available.
  """
  @spec create_function(
          function_name :: binary,
          handler :: binary,
          zipfile :: binary
        ) :: ExAws.Operation.JSON.t()
  @spec create_function(
          function_name :: binary,
          handler :: binary,
          zipfile :: binary,
          opts :: create_function_opts
        ) :: ExAws.Operation.JSON.t()
  def create_function(function_name, handler, zipfile, opts \\ []) do
    data =
      opts
      |> normalize_opts
      |> Map.merge(%{
        "FunctionName" => function_name,
        "Handler" => handler,
        "Code" => %{"ZipFile" => zipfile}
      })

    request(:create_function, data, "/2015-03-31/functions")
  end

  @doc """
  Delete an event source mapping.
  """
  @spec delete_event_source_mapping(source_mapping_uuid :: binary) :: ExAws.Operation.JSON.t()
  def delete_event_source_mapping(source_mapping_uuid) do
    request(
      :delete_event_source_mapping,
      %{},
      "/2015-03-31/event-source-mappings/#{source_mapping_uuid}"
    )
  end

  @doc """
  Delete a lambda function.
  """
  @spec delete_function(function_name :: binary) :: ExAws.Operation.JSON.t()
  def delete_function(function_name) do
    request(:delete_function, %{}, "/2015-03-31/functions/#{function_name}")
  end

  @doc """
  Get an event source mapping.
  """
  @spec get_event_source_mapping(source_mapping_uuid :: binary) :: ExAws.Operation.JSON.t()
  def get_event_source_mapping(source_mapping_uuid) do
    request(
      :get_event_source_mapping,
      %{},
      "/2015-03-31/event-source-mappings/#{source_mapping_uuid}"
    )
  end

  @doc """
  Get a function.
  """
  @spec get_function(function_name :: binary) :: ExAws.Operation.JSON.t()
  def get_function(function_name) do
    request(:get_function, %{}, "/2015-03-31/functions/#{function_name}/versions/HEAD")
  end

  @doc """
  Get a function configuration.
  """
  @spec get_function_configuration(function_name :: binary) :: ExAws.Operation.JSON.t()
  def get_function_configuration(function_name) do
    request(
      :get_function_configuration,
      %{},
      "/2015-03-31/functions/#{function_name}/versions/HEAD/configuration"
    )
  end

  @doc """
  Get a function access policy.
  """
  @spec get_policy(function_name :: binary) :: ExAws.Operation.JSON.t()
  def get_policy(function_name) do
    request(:get_policy, %{}, "/2015-03-31/functions/#{function_name}/versions/HEAD/policy")
  end

  @doc """
  Invoke a lambda function.
  """
  @type invoke_opts :: [
          {:invocation_type, :event | :request_response | :dry_run}
          | {:log_type, :none | :tail}
          | {:qualifier, String.t()}
          | {:parser, fun()}
        ]
  @spec invoke(function_name :: binary, payload :: map(), client_context :: map()) ::
          ExAws.Operation.JSON.t()
  @spec invoke(
          function_name :: binary,
          payload :: map(),
          client_context :: map(),
          opts :: invoke_opts
        ) :: ExAws.Operation.JSON.t() | ExAws.Operation.RestQuery.t()
  @context_header "X-Amz-Client-Context"
  def invoke(function_name, payload, client_context, opts \\ []) do
    {qualifier, opts} = Map.pop(Enum.into(opts, %{}), :qualifier)

    headers =
      [
        invocation_type: "X-Amz-Invocation-Type",
        log_type: "X-Amz-Log-Type",
        xray_trace_id: "X-Amzn-Trace-Id"
      ]
      |> Enum.reduce([], fn {opt, header}, headers ->
        case Map.fetch(opts, opt) do
          :error -> headers
          {:ok, nil} -> headers
          {:ok, value} -> [{header, value |> camelize_key} | headers]
        end
      end)

    headers =
      case client_context do
        %{} ->
          headers

        context ->
          [{@context_header, context} | headers]
      end

    url = "/2015-03-31/functions/#{function_name}/invocations"
    url = if qualifier, do: url <> "?Qualifier=#{qualifier}", else: url

    request(
      :invoke,
      payload,
      url,
      [],
      headers,
      fn operation, config ->
        headers =
          operation.headers
          |> List.keyfind(@context_header, 0, nil)
          |> case do
            nil ->
              operation.headers

            {_, val} ->
              new_val = val |> config.json_codec.encode! |> Base.encode64()
              List.keyreplace(operation.headers, @context_header, 0, {@context_header, new_val})
          end

        %{operation | headers: headers}
      end,
      Map.get(opts, :parser, nil)
    )
  end

  @doc """
  Invoke a lambda function asynchronously.
  """
  @spec invoke_async(function_name :: binary, args :: map()) :: ExAws.Operation.JSON.t()
  def invoke_async(function_name, args) do
    Logger.info(
      "This API is deprecated. See invoke/5 with the Event value set as invocation type"
    )

    request(
      :invoke,
      args |> normalize_opts,
      "/2014-11-13/functions/#{function_name}/invoke-async/"
    )
  end

  @doc """
  List event source mappings.
  """
  @type list_event_source_mappings_opts :: [
          {:event_source_arn, binary}
          | {:function_name, binary}
          | {:marker, binary}
          | {:max_items, pos_integer}
        ]
  @spec list_event_source_mappings() :: ExAws.Operation.JSON.t()
  @spec list_event_source_mappings(opts :: list_event_source_mappings_opts) ::
          ExAws.Operation.JSON.t()
  def list_event_source_mappings(opts \\ []) do
    params =
      opts
      |> normalize_opts

    request(:list_event_source_mappings, %{}, "/2015-03-31/event-source-mappings/", params)
  end

  @doc """
  List functions.
  """
  @type list_functions_opts :: [
          {:marker, binary}
          | {:max_items, pos_integer}
        ]
  @spec list_functions() :: ExAws.Operation.JSON.t()
  @spec list_functions(opts :: list_functions_opts) :: ExAws.Operation.JSON.t()
  def list_functions(opts \\ []) do
    request(:list_functions, %{}, "/2015-03-31/functions/", normalize_opts(opts))
  end

  @doc """
  Remove individual permissions from an function's access policy.
  """
  @spec remove_permission(function_name :: binary, statement_id :: binary) ::
          ExAws.Operation.JSON.t()
  def remove_permission(function_name, statement_id) do
    request(
      :remove_permission,
      %{},
      "/2015-03-31/functions/#{function_name}/versions/HEAD/policy/#{statement_id}"
    )
  end

  @doc """
  Update event source mapping.
  """
  @spec update_event_source_mapping(uuid :: binary, attrs_to_update :: map()) ::
          ExAws.Operation.JSON.t()
  def update_event_source_mapping(uuid, attrs_to_update) do
    request(
      :update_event_source_mapping,
      attrs_to_update,
      "/2015-03-31/event-source-mappings/#{uuid}"
    )
  end

  @doc """
  Update function code.
  """
  @spec update_function_code(function_name :: binary, zipfile :: binary) ::
          ExAws.Operation.JSON.t()
  def update_function_code(function_name, zipfile) do
    data = %{"ZipFile" => zipfile}

    request(
      :update_function_code,
      data,
      "/2015-03-31/functions/#{function_name}/versions/HEAD/code"
    )
  end

  @doc """
  Update a function configuration.
  """
  @spec update_function_configuration(function_name :: binary, configuration :: map()) ::
          ExAws.Operation.JSON.t()
  def update_function_configuration(function_name, configuration) do
    data = configuration |> normalize_opts

    request(
      :update_function_configuration,
      data,
      "/2015-03-31/functions/#{function_name}/configuration"
    )
  end

  @doc """
  Tag the Lambda resource.
  """
  @spec tag_resource(function_name :: binary, tags :: map()) :: ExAws.Operation.JSON.t()
  def tag_resource(function_arn, tags) do
    data = %{"Tags" => Map.new(tags, fn {k, v} -> {to_string(k), v} end)}

    request(
      :tag_resource,
      data,
      "/2017-03-31/tags/#{function_arn}"
    )
  end

  defp normalize_opts(opts) do
    opts
    |> Map.new()
    |> camelize_keys
  end

  defp request(
         action,
         data,
         path,
         params \\ [],
         headers \\ [],
         before_request \\ nil,
         parser \\ nil
       )

  defp request(action, data, path, params, _headers, _before_request, parser)
       when is_function(parser, 2) and not is_nil(parser) do
    path = [path, "?", params |> URI.encode_query()] |> IO.iodata_to_binary()
    http_method = @actions |> Map.fetch!(action)

    %ExAws.Operation.RestQuery{
      service: :lambda,
      http_method: http_method,
      path: path,
      body: data,
      parser: parser
    }
  end

  defp request(action, data, path, params, headers, before_request, _) do
    path = [path, "?", params |> URI.encode_query()] |> IO.iodata_to_binary()
    http_method = @actions |> Map.fetch!(action)

    ExAws.Operation.JSON.new(:lambda, %{
      http_method: http_method,
      path: path,
      data: data,
      headers: [{"content-type", "application/json"} | headers],
      before_request: before_request
    })
  end
end
